File: AlphaStream/Models/Order.py
from .OrderEnums import *
from .OrderEvent import *
from datetime import datetime


class Order:
    def __init__(self, json):

        # Unique hash of the algorithm
        self.AlgorithmId = json.get('algorithm-id')

        # Unique hash of the order
        self.OrderId = json.get('order-id')

        # Unique hash of the source of the order
        self.Id = f"{self.AlgorithmId}-{self.OrderId}"

        # QuantConnect identifier of the order asset
        self.Symbol = json.get('symbol')

        # Id of any orders to be filled before triggering this order
        self.ContingentId = json.get('contingent-id', 0)

        # Id of the brokerage the order was placed with
        self.BrokerId = json.get('broker-id', [])

        # Order fill price
        self.Price = json.get('price')

        # Currency of the order price
        self.PriceCurrency = json.get('price-currency')

        # Unix time when the order was created
        self.CreatedTime = datetime.utcfromtimestamp(json.get('created-time'))

        # Unix time the order was last filled (fully or partially)
        self.LastFillTime = datetime.utcfromtimestamp(json.get('last-fill-time')) if json.get('last-fill-time') is not None else None

        # Unix time the order was last updated
        self.LastUpdateTime = datetime.utcfromtimestamp(json.get('last-update-time')) if json.get('last-update-time') is not None else None

        # Unix time the order was canceled
        self.CanceledTime = datetime.utcfromtimestamp(json.get('canceled-time')) if json.get('canceled-time') is not None else None

        # Number of shares of the order
        self.Quantity = json.get('quantity')

        # Order type - market, market on open, market on close, limit, stop limit, stop market, option exercise
        self.Type = OrderType(json.get('type'))

        # Fill-status of the order - new, submitted, partially filled, filled, canceled, none, invalid, cancel pending, update submitted
        self.Status = OrderStatus(json.get('status'))

        # Message attached to the order
        self.Tag = json.get('tag')

        # Order direction - buy, sell, hold
        self.Direction = OrderDirection(json.get('direction'))

        # Mid-price of the security when the order was submitted
        self.SubmissionLastPrice = json.get('submission-last-price')

        # Ask price of the security when the order was submitted
        self.SubmissionAskPrice = json.get('submission-ask-price')

        # Bid price of the security when the order was submitted
        self.SubmissionBidPrice = json.get('submission-bid-price')

        # Stop-price set for the order
        self.StopPrice = json.get('stop-price')

        # Boolean if the stop price was hit
        self.StopTriggered = json.get('stop-triggered')

        # Limit-price set for the order
        self.LimitPrice = json.get('limit-price')

        # Type of order fill duration
        self.TimeInForceType = json.get('time-in-force-type')

        # Unix time when the order time-in-force ends
        self.TimeInForceExpiry = datetime.utcfromtimestamp(json.get('time-in-force-expiry')) if json.get('time-in-force-expiry') is not None else None

        # Where the order was placed - in sample, out of sample, live trading
        self.Source = json.get('source')

        # List of OrderEvents associated with the order
        self.OrderEvents = [OrderEvent(x) for x in json.get('events', [])]

    def __repr__(self):
        rep = f'ID: {self.Id} Source: "{self.Source}" Symbol: {self.Symbol} Status: {self.Status.name} CreatedTime: {self.CreatedTime} Direction: {self.Direction.name} Quantity: {self.Quantity} Type: {self.Type.name} TimeInForceType: {self.TimeInForceType}'

        if self.TimeInForceExpiry is not None:
            rep += f' TimeInForceExpiry: {self.TimeInForceExpiry}'

        if len(self.BrokerId) > 0:
            rep += f' BrokerID: {",".join(self.BrokerId)}'

        if self.Price != 0:
            rep += f' Price: {self.Price} {self.PriceCurrency}'

        if self.ContingentId != 0:
            rep += f' ContingentId: {self.ContingentId}'

        if self.LastUpdateTime is not None:
            rep += f' LastUpdateTime: {self.LastUpdateTime}'

        if self.LastFillTime is not None:
            rep += f' LastFillTime: {self.LastFillTime}'

        if self.CanceledTime is not None:
            rep += f' CanceledTime: {self.CanceledTime}'

        if self.LimitPrice is not None:
            rep += f' LimitPrice: {self.LimitPrice}'

        if self.StopPrice is not None:
            rep += f' StopPrice: {self.StopPrice}'

        if self.SubmissionLastPrice != 0:
            rep += f' SubmissionLastPrice: {self.SubmissionLastPrice}'

        if self.SubmissionAskPrice != 0:
            rep += f' SubmissionAskPrice: {self.SubmissionAskPrice}'

        if self.SubmissionBidPrice != 0:
            rep += f' SubmissionBidPrice: {self.SubmissionBidPrice}'

        if self.Tag is not None:
            rep += f' Tag: {self.Tag}'

        if len(self.OrderEvents) > 0:
            rep += ' OrderEvents: [{' + "},{".join([x.__repr__(extended = False) for x in self.OrderEvents]) + '}]'

        return rep

File: QuantConnect.AlphaStream/Models/Orders/Order.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Text;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using QuantConnect.Util;

namespace QuantConnect.AlphaStream.Models.Orders
{
    public class Order
    {
        /// <summary>
        /// The unique order id
        /// </summary>
        [JsonProperty("id")]
        public string Id => $"{AlgorithmId}-{OrderId}";

        /// <summary>
        /// Algorithm Id, BacktestId or DeployId
        /// </summary>
        [JsonProperty("algorithm-id")]
        public string AlgorithmId { get; set; }

        /// <summary>
        /// Order ID
        /// </summary>
        [JsonProperty("order-id")]
        public int OrderId { get; set; }

        /// <summary>
        /// Order id to process before processing this order.
        /// </summary>
        [JsonProperty("contingent-id")]
        public int ContingentId { get; set; }

        /// <summary>
        /// Brokerage Id for this order for when the brokerage splits orders into multiple pieces
        /// </summary>
        [JsonProperty("broker-id")]
        public List<string> BrokerId { get; set; } = new List<string>();

        /// <summary>
        /// Symbol of the Asset
        /// </summary>
        [JsonProperty("symbol")]
        public string Symbol { get; set; }

        /// <summary>
        /// Price of the Order.
        /// </summary>
        [JsonProperty("price")]
        public decimal Price { get; set; }

        /// <summary>
        /// Currency for the order price
        /// </summary>
        [JsonProperty("price-currency")]
        public string PriceCurrency { get; set; }

        /// <summary>
        /// Gets the utc time this order was created. Alias for <see cref="Time"/>
        /// </summary>
        [JsonProperty("created-time"), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime CreatedTime { get; set; }

        /// <summary>
        /// Gets the utc time the last fill was received, or null if no fills have been received
        /// </summary>
        [JsonProperty("last-fill-time", NullValueHandling = NullValueHandling.Ignore), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime? LastFillTime { get; set; }

        /// <summary>
        /// Gets the utc time this order was last updated, or null if the order has not been updated.
        /// </summary>
        [JsonProperty("last-update-time", NullValueHandling = NullValueHandling.Ignore), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime? LastUpdateTime { get; set; }

        /// <summary>
        /// Gets the utc time this order was canceled, or null if the order was not canceled.
        /// </summary>
        [JsonProperty("canceled-time", NullValueHandling = NullValueHandling.Ignore), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime? CanceledTime { get; set; }

        /// <summary>
        /// Number of shares to execute.
        /// </summary>
        [JsonProperty("quantity")]
        public decimal Quantity { get; set; }

        /// <summary>
        /// Order Type
        /// </summary>
        [JsonProperty("type"), JsonConverter(typeof(StringEnumConverter), true)]
        public OrderType Type { get; set; }

        /// <summary>
        /// Status of the Order
        /// </summary>
        [JsonProperty("status"), JsonConverter(typeof(StringEnumConverter), true)]
        public OrderStatus Status { get; set; }

        /// <summary>
        /// Tag the order with some custom data
        /// </summary>
        [DefaultValue(""), JsonProperty("tag", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public string Tag { get; set; }

        /// <summary>
        /// Order Direction Property based off Quantity.
        /// </summary>
        [JsonProperty("direction"), JsonConverter(typeof(StringEnumConverter), true)]
        public OrderDirection Direction { get; set; }

        /// <summary>
        /// The current price at order submission time
        /// </summary>
        [JsonProperty("submission-last-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal SubmissionLastPrice { get; set; }

        /// <summary>
        /// The ask price at order submission time
        /// </summary>
        [JsonProperty("submission-ask-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal SubmissionAskPrice { get; set; }

        /// <summary>
        /// The bid price at order submission time
        /// </summary>
        [JsonProperty("submission-bid-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal SubmissionBidPrice { get; set; }

        /// <summary>
        /// The current stop price
        /// </summary>
        [JsonProperty("stop-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal? StopPrice { get; set; }

        /// <summary>
        /// Signal showing the "StopLimitOrder" has been converted into a Limit Order
        /// </summary>
        [JsonProperty("stop-triggered", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public bool? StopTriggered { get; set; }

        /// <summary>
        /// The current limit price
        /// </summary>
        [JsonProperty("limit-price", DefaultValueHandling = DefaultValueHandling.Ignore)]
        public decimal? LimitPrice { get; set; }

        /// <summary>
        /// The time in force type
        /// </summary>
        [JsonProperty("time-in-force-type")]
        public string TimeInForceType { get; set; }

        /// <summary>
        /// The time in force expiration time if any
        /// </summary>
        [JsonProperty("time-in-force-expiry", DefaultValueHandling = DefaultValueHandling.Ignore), JsonConverter(typeof(DoubleUnixSecondsDateTimeJsonConverter))]
        public DateTime? TimeInForceExpiry { get; set; }

        /// <summary>
        /// Specifies the source of this order, live trading/backtesting/out of sample
        /// </summary>
        [JsonProperty("source")]
        public string Source { get; set; }

        /// <summary>
        /// The orders <see cref="OrderEvent"/>
        /// </summary>
        [JsonProperty("events")]
        public List<OrderEvent> OrderEvents { get; set; } = new List<OrderEvent>();

        /// <summary>
        /// Returns a string that represents the current object.
        /// </summary>
        public override string ToString()
        {
            var stringBuilder = new StringBuilder();

            stringBuilder.Append($"ID: {Id} Source '{Source}' Symbol: {Symbol} Status: {Status} CreatedTime: {CreatedTime} Direction {Direction} Quantity {Quantity} Type: {Type} TimeInForceType: {TimeInForceType}");

            if (TimeInForceExpiry.HasValue)
            {
                stringBuilder.Append($" TimeInForceExpiry: {TimeInForceExpiry}");
            }

            if (BrokerId.Count > 0)
            {
                stringBuilder.Append($" BrokerId: [{string.Join(",", BrokerId)}]");
            }

            if (Price != 0)
            {
                stringBuilder.Append($" Price: {Price} {PriceCurrency}");
            }

            if (ContingentId != 0)
            {
                stringBuilder.Append($" ContingentId: {ContingentId}");
            }

            if (LastUpdateTime.HasValue)
            {
                stringBuilder.Append($" LastUpdateTime: {LastUpdateTime}");
            }
            if (LastFillTime.HasValue)
            {
                stringBuilder.Append($" LastFillTime: {LastFillTime}");
            }
            if (CanceledTime.HasValue)
            {
                stringBuilder.Append($" CanceledTime: {CanceledTime}");
            }

            if (LimitPrice.HasValue)
            {
                stringBuilder.Append($" LimitPrice: {LimitPrice.Value}");
            }
            if (StopPrice.HasValue)
            {
                stringBuilder.Append($" StopPrice: {StopPrice.Value}");
            }

            if (SubmissionLastPrice != 0)
            {
                stringBuilder.Append($" SubmissionLastPrice: {SubmissionLastPrice}");
            }
            if (SubmissionAskPrice != 0)
            {
                stringBuilder.Append($" SubmissionAskPrice: {SubmissionAskPrice}");
            }
            if (SubmissionBidPrice != 0)
            {
                stringBuilder.Append($" SubmissionBidPrice: {SubmissionBidPrice}");
            }

            if (!string.IsNullOrEmpty(Tag))
            {
                stringBuilder.Append($" Tag: '{Tag}'");
            }

            if (OrderEvents.Count > 0)
            {
                stringBuilder.Append(" OrderEvents: [{");
                stringBuilder.Append(string.Join("},{", OrderEvents.Select(orderEvent => orderEvent.ToString(extended:false))));
                stringBuilder.Append("}]");
            }

            return stringBuilder.ToString();
        }
    }
}
